jax/lax.py

```diff
@@ -2383,33 +2388,82 @@ def select_and_scatter_add_transpose(
     select_and_scatter_add_transpose
 
 
-def select_and_gather_add_shape_rule(
-    tangents, operand, select_prim, window_dimensions, window_strides, padding):
+def _select_and_gather_add_shape_rule(
+    tangents, operand, select_prim, pair_select_jaxpr, pair_select_consts,
+    window_dimensions, window_strides, padding):
   if tangents.shape != operand.shape:
     msg = ("select_and_gather_add tangents and operand shapes must match, "
            "got {} and {}.")
     raise TypeError(msg.format(tangents.shape, operand.shape))
   return common_reduce_window_shape_rule(operand, window_dimensions,
                                          window_strides, padding)
 
-def select_and_gather_add_translation(
-    c, tangents, operand, select_prim, window_dimensions, window_strides,
-    padding):
-  raise NotImplementedError("No efficient translation.")
 
-def select_and_gather_add_transpose(
-    t, tangents, operand, select_prim, window_dimensions, window_strides,
-    padding):
+_UINT_DTYPES = {
+  16: onp.uint16,
+  32: onp.uint32,
+  64: onp.uint64,
+}
+
+def _select_and_gather_add_pair_dtype(dtype):
+  bits = onp.finfo(dtype).bits
+  return _UINT_DTYPES[bits * 2]
+
+def _select_and_gather_add_pair_reducer(x, y, dtype=None, select_prim=None):
+  bits = _dtype(x).type(onp.finfo(dtype).bits)
+  uint_dtype = _UINT_DTYPES[bits]
+  sx = convert_element_type(shift_right_logical(x, bits), uint_dtype)
+  sx = bitcast_convert_type(sx, dtype)
+  sy = convert_element_type(shift_right_logical(y, bits), uint_dtype)
+  sy = bitcast_convert_type(sy, dtype)
+  return select(select_prim.bind(sx, sy), x, y)
+
+def _select_and_gather_add_translation(
+    c, tangents, operand, select_prim, pair_select_jaxpr, pair_select_consts,
+    window_dimensions, window_strides, padding):
+  # XLA doesn't yet implement ReduceWindow on tuples (Google bug b/73062247), so
+  # we implement a pair-wise ReduceWindow by packing two k-bit values into
+  # 2k-bit unsigned integer using bit tricks. This will only work for 32-bit
+  # inputs, and furthermore it also requires jax_enable_x64 to be set.
```

Comment:
For now, maybe just add a check that `jax_enable_x64 == True` here?

Reply:
Done.